### DISCUSSION TITLE
Using edge features with NeighborSampler and SageConv

### DISCUSSION THREAD

**Discussion by a-ism256** [pyg-team/pytorch_geometric#2602] [Q&A] [answered]
Hello,

I'm trying to make NeighborSampler and SageConv work for my dataset (node classification) which has both node and edge features. Each graph is a complete graph with no self loops.

So far I'm trying to use the SageConv modification proposed [here](https://github.com/rusty1s/pytorch_geometric/issues/1282) which was further implemented by the person asking the question [here](https://github.com/kkonevets/geo_detection/blob/9421a591123c380a1f232b6bff598cae8ff29a23/sage_conv.py).

However, I'm running into issues using the NeighborSampler with my data for the following reasons:
1. How can I properly use e_id to get the corresponding edge features? Does e_id correspond to the index of the corresponding edge in the original edge_index or to the actual id of the edge (I don't understand the latter case since an edge in edge_index is defined by a pair of node ids).
2. My dataset has multiple graphs and NeighborSampler does not seem to be able to sample from multiple graphs. So far, the only solution I see is looping over the graphs in my dataset and creating a neighbor sampler for each. Or loading everything into a single Data graph however only nodes originally belonging to the same graph would be connected.

I would highly appreciate if someone could help me figure these out.

**Comment by wsad1** [comment, marked as answer]
1. Yes e_id corresponds to the index of the edge in the full(original) graph. So `edge_attr[ed_id]` should give you the edge attributes for the edges in the sample. Check [this](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.NeighborSampler) for more documentation and examples..
2. Yes, creating on large graph with multiple disconnected components as described [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html) seems like the best option.

**Reply by a-ism256** [reply]
Hello,

Thanks for answering my questions.

Just to make sure I understand your point on the second question, I should batch all my graphs into a single graph by using `DataLoader` and setting the batch size to the total number of graphs and then use the `NeighborSampler` on the resulting batch? An example would be the following:

`loader = DataLoader(dataset, batch_size=len(dataset), shuffle=True)`
`batch = list(loader)[0]`
`sampler = NeighborSampler(batch.edge_index, node_idx=batch.train_mask,
                                           sizes=nbr_sizes, batch_size=1024,
                                           shuffle=True, num_workers=12)`

Also, regarding the first point, does the graph need to have self loops for `NeighborSampler` to use `e_id` properly as indicated [here](https://github.com/rusty1s/pytorch_geometric/issues/64#issuecomment-523634240) and if so why?

**Reply by rusty1s** [reply]
Yes, this looks good. Self-loops are not necessary anymore :)

**Reply by a-ism256** [reply]
Hello again!

While the above method seemed to work I ran into an OOM error since the single giant graph turned out to be too big. Therefore I don't think I can set the `batch_size` to `len(dataset)` but will have to rely on mini-batches.

My question is how would the `NeighborSampler` be used in this case? Should I create a sampler for each batch as follow?

```
for batch in loader:
    sampler = NeighborSampler(batch.edge_index, node_idx=batch.train_mask, sizes=nbr_sizes, batch_size=1024, shuffle=True, num_workers=12)
    for batch_size, n_id, adjs in sampler:
        ...
```

I'm worried about using `NeighborSampler` in this case and ending up sampling across different graphs.

Honestly at this point I'm wondering if there could be some other framework to suit my application: inductive node classification using node and edge features with a multi-graph dataset. Since I'm trying to do this using SAGEConv I'm having to do a lot of workarounds and I'm starting to go in circles.

Appreciate and advice you might provide.

**Reply by rusty1s** [reply]
How big is your total dataset? And how big is each individual graph? Are you talking about CPU or GPU OOM errors? Since you are already holding `dataset` in CPU memory, creating a single giant graph from it should fit in CPU memory as well.

**Reply by a-ism256** [reply]
It's a CPU OOM issue. I actually ran into the OOM error when I was trying to use the `InMemoryDataset` so I switched to `Dataset` where each graph is saved as an individual file to disk.

However, if I were still to do what I mentioned [above](https://github.com/rusty1s/pytorch_geometric/discussions/2602#discussioncomment-762871) loading all the individual graphs into one giant graph by setting `batch_size = len(dataset)` would again cause the OOM error.

**Reply by rusty1s** [reply]
I see. I think your best bet is to create a sampler for each individual graph object, and iterate over each sampler for each graph.
